plugins/autoip.rb

```diff
@@ -35,29 +36,104 @@ def self.pre_validate options
       # Remove any 'DOMAIN\' prefix from the username
       username = options[:username]
       username = username.gsub(/^.+\\(.*)/, '\1')
-      puts "Requesting IP in #{options[:subnet]} vlan"
+      if (not options[:password])
+        print 'Password: '
+        options[:password] = STDIN.noecho(&:gets).chomp
+        puts ''
+      end
+
+
+      # Authenticate and get a token for further operations
+
+      puts "Requesting auth token for #{username}"
+      uri = options[:add_uri].gsub(/APIAPP/, {
+        'APIAPP' => options[:apiapp],
+      })
+      uri = uri 
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("user/"))
+      request = Net::HTTP::Post.new(uri.request_uri)
+      request.basic_auth username, options[:password]
+      retries = [3, 5, 10]
+      begin
+        response = Net::HTTP.start(uri.host, uri.port, :read_timeout => 5, :use_ssl => true){|http| http.request(request)}
+        rescue Net::ReadTimeout
+          if delay = retries.shift
+            sleep delay
+            retry
+          else
+            abort "The request has timed out, check your username/password and try again"
+          end
+        end
+      if response.code != "200"
+        abort "There was an error requesting your IP address, IPAM returned code: #{response.code}, #{JSON.parse(response.body)["message"]}"
+      end
+      auth_token = JSON.parse(response.body)["data"]["token"]
+   
+      # Check for an existing host in IPAM
+
+      puts "Checking for existing IP for host #{options[:hostname]}"
+      uri = options[:add_uri]
+      uri = URI.escape(uri)
+      uri = URI.parse(uri.concat("addresses/search_hostname/#{options[:hostname]}/"))
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl = true
+      request = Net::HTTP::Get.new(uri.request_uri)
+      request.add_field("token",  auth_token)
+      response = http.request(request)
+      if response.code == "404"
+        puts "Requesting IP for #{options[:hostname]}"
+      elsif response.code == "200"
+        abort "#{options[:hostname]} is already assigned #{JSON.parse(response.body)["data"][0]["ip"]}"
```

Comment:
By aborting here, you prevent rebuilding a host don't you?  If we delete the VM, but not the host record, we should still be able to rebuild using the original IP, I should think.